src/Models/NonhydrostaticModels/pressure_correction.jl

```diff
@@ -1,21 +1,57 @@
 import Oceananigans.TimeSteppers: calculate_pressure_correction!, pressure_correct_velocities!
 
+const c = Center()
+const f = Face()
+
+using Oceananigans.Grids: node
+using Oceananigans.StokesDrifts: StokesDrift, parameters_tuple
+
+function modulate_by_stokes_drift!(model, sgn, t=time(model))
+    stokes_drift = model.stokes_drift
+    if stokes_drift isa StokesDrift
+        grid = model.grid
+        arch = architecture(grid)
+        u, v, w = model.velocities
+        launch!(arch, grid, :xyz, _modulate_by_stokes_drift, u, v, w, sgn, grid, t, stokes_drift)
+    end
+    return nothing
+end
+
+@kernel function _modulate_by_stokes_drift(u, v, w, sgn, grid, time, sd)
+    i, j, k = @index(Global, NTuple)
+
+    pt = parameters_tuple(sd)
+    Xu = node(i, j, k, grid, f, c, c)
+    Xv = node(i, j, k, grid, c, f, c)
+    Xw = node(i, j, k, grid, c, c, f)
+
+    @inbounds begin
+        u[i, j, k] = u[i, j, k] + sgn * sd.uˢ(Xu..., time, pt...)
+        v[i, j, k] = v[i, j, k] + sgn * sd.vˢ(Xv..., time, pt...)
+        w[i, j, k] = w[i, j, k] + sgn * sd.wˢ(Xw..., time, pt...)
```

Comment:
Is this subtracting/adding the Stokes drift at a specific depth from a depth-averaged current? Perhaps this is the kind of issue you were alluding to @qingli411

Reply:
Yes, I think here we should use cell-averaged Stokes drift and in the finite volume approximation of $\partial_z u^S$ we will need the values at specific locations. When Stokes drift is specified as an analytical function, we can make sure this is done consistently. But what I'm not sure about is that, if the Stokes drift cannot be specified as an analytical function, what values should we pass in? So maybe also keep the option to pass in the derivatives of Stokes drift?

Reply:
Hmm, it's a good idea to discuss the discretization. I think on a C-grid we probably want to have the Stokes drift at the same location as the velocity components. This would not be cell centers, but Face, Center, Center for u^S, Center, Face, Center for v^S, etc.

Probably, there is a correct discretization of the term $\tilde \omega \times u^L$ that conserves energy (where $\tilde \omega$ is the curl of the Stokes drift, similar to how there is a correct discretization of vorticity / Coriolis that conserves energy.

For example see how we do it for Coriolis on the sphere:

https://github.com/CliMA/Oceananigans.jl/blob/d8273e7a07e8f53f9772f87f67f6eca8d17cb98c/src/Coriolis/hydrostatic_spherical_coriolis.jl#L112-L119

If we want to be able to naturally compute $\omega$ at the "C-grid vorticity location" then I believe we want u^S at the same location as the velocity field.

Reply:
I think we can build a simple example to demonstrate conservation of energy (which is $(u^L)^2 / 2$) which will show that we've done the discretization correctly.

Reply:
> But what I'm not sure about is that, if the Stokes drift cannot be specified as an analytical function, what values should we pass in? So maybe also keep the option to pass in the derivatives of Stokes drift?

That's a fair point. Though it would still be possible to compute the Stokes drift numerically by integrating its derivatives, but maybe that's not so convenient.

Reply:
> > But what I'm not sure about is that, if the Stokes drift cannot be specified as an analytical function, what values should we pass in? So maybe also keep the option to pass in the derivatives of Stokes drift?
> 
> That's a fair point. Though it would still be possible to compute the Stokes drift numerically by integrating its derivatives, but maybe that's not so convenient.

The pressure solver depends only on the divergence of the velocity field. If the Stokes drift gradients were provided, would we now need to specify every derivative since the terms needed for the pressure solver Eulerian-ization ($\partial_xu^s$, $\partial_yv^s$ and $\partial_zw^s$) are distinct from the ones needed for the wave-averaged forces: 

https://github.com/CliMA/Oceananigans.jl/blob/0fc5708f0c4dd953f2433b8db2ff22f2dbab2639/src/StokesDrifts.jl#L153-L167

Reply:
OK. Maybe asking the users to specify every derivative is too complicated. I agree that the easiest approach is to specify the Stokes drift as $u^S(x,y,z,t)$ and compute the required values and derivatives internally. The difference between the Stokes drift at a specific depth and a depth-averaged Stokes drift around that depth should be small if we have sufficient resolution.

Reply:
> Maybe asking the users to specify every derivative is too complicated.

It's what we do currently 😂 

We don't _have_ to choose to do anything specific. We can keep the existing structure, and add a new one wherein users specify the Stokes drift only. Then the test of time shall arbitrate.

> The difference between the Stokes drift at a specific depth and a depth-averaged Stokes drift around that depth should be small if we have sufficient resolution.

I do think that either way, sufficient resolution renders either discretization similar. But it would be interesting to understand the implications of inadequate resolution too.

Reply:
I think as @BrodiePearson pointed out we will need to specify three more spatial derivatives ($\partial_x u^s$, $\partial_y v^s$, and $\partial_z w^s$)...

I agree. I was not saying we should ignore the differences. But I cannot think of a clean way to specify the Stokes drift. I think we need both the Stokes drift at a specific depth and the depth-averaged Stokes drift around that depth. Maybe asking users to specify both is another option? I was also thinking of an option to specify the wave spectrum, or a simplified one with less frequency bins as Brandon did in MOM6. But maybe these different options can be wrapped into another layer of interface in the StokesDrift module, or defined by users. Internally in Oceananigans, I guess we need $u^s$ and $v^s$ defined as depth-averaged values at the locations of $u$ and $v$, $w^s$ defined at the location of $w$, spatial derivatives should also be depth-averaged over a grid cell?

Reply:
We can compute the finite volume derivatives of the Stokes drift field in the same way that we compute the discrete finite volume derivatives of the prognostic velocity field. The simplest stencil is second-order accurate (it looks just like a finite difference stencil). You could be right though that the derivatives may be known with higher accuracy if users are allowed to supply them.

It wouldn't be too much work to put an interface where the needed derivatives may be optionally provided, and otherwise are computed with a finite volume method.